### PR TITLE
KZG point eval precompile

### LINKED ISSUE #4822: KZG point evaluation precompile
Leverage the work done by the KZG library being used by Teku: https://github.com/ConsenSys/jc-kzg-4844 from a precompiled contract as specified here:

https://eips.ethereum.org/EIPS/eip-4844#point-evaluation-precompile

- add dependency on kzg lib above. Figure out how to properly package native code for inclusion in all distros.
- statics for new precompile reserved address
- add to mainnet precompile registry - add to Shanghai, or add another for Cancun?
- truth table style unit tests to validate math is correct - where do we source that from? BLS12 tests have .csv to go with it for validation.
- higher level tests at the EVM to cover kzg assertion failures are handled properly
- command line args to override trusted setup file, and documentation for that
- package the mainnet kzg trusted setup file so it is delivered via distribution in such a way that it can be loaded by native lib from the filesystem; loading from classpath won't work.
- add support for reftests that cover it

### PR DESCRIPTION
**Pull request by jflo**
## PR description

Adds a new precompile as part of EIP-4844 which allows contracts to evaluate KZG commitments. 

## Fixed Issue(s)
fixes #4822 

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,14 +6,18 @@
...
 ### Bug Fixes
 
 ## 23.1.0-beta
 
 ### Breaking Changes
 - GoQuorum-compatible privacy is deprecated and will be removed in 23.4
 - IBFT 1.0 is deprecated and will be removed in 23.4
+
+
+## 22.10.4
```
Comment (siladu) [inline CHANGELOG.md:20]: another bad merge?

Comment (jflo) [reply CHANGELOG.md:20]: indeed. converting back to draft till I can sort this tomorrow.

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java [hunk 2]
```diff
@@ -86,4 +88,11 @@ public static void populateForBLS12(
     registry.put(Address.BLS12_MAP_FP_TO_G1, new BLS12MapFpToG1PrecompiledContract());
     registry.put(Address.BLS12_MAP_FP2_TO_G2, new BLS12MapFp2ToG2PrecompiledContract());
   }
+
+  public static void populateForCancun(
+      final PrecompileContractRegistry registry, final GasCalculator gasCalculator) {
+    populateForIstanbul(registry, gasCalculator);
```
Comment (siladu) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java:94]: Why is this not populateForBLS12?

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java [hunk 3]
```diff
@@ -498,7 +498,7 @@ public void toURI_WithHostnameShouldWorkWhenDnsEnabledAndUpdateEnabled()
         "enode://"
             + VALID_NODE_ID
             + "@"
-            + InetAddress.getLocalHost().getHostName()
+            + InetAddress.getLocalHost().getCanonicalHostName()
```
Comment (siladu) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/peers/EnodeURLImplTest.java:501]: ooo is this a fix for when this fails locally?

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      }
+      CKZG4844JNI.loadNativeLibrary(bitLength);
+      try {
+        CKZG4844JNI.loadTrustedSetup(pathToSetup);
+      } catch (RuntimeException alreadyLoaded) {
+        if (alreadyLoaded.getMessage().contains("Trusted Setup is already loaded")) {
+        } else {
+          throw alreadyLoaded;
+        }
+      }
```
Comment (siladu) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:76]: what's going on here? Is there not an exception type for 'already loaded'?

File: evm/src/test/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompileContractTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    PrecompiledContract.PrecompileContractResult result = contract.computePrecompile(input, toRun);
+
+    MessageFrame.State endState = result.getState();
+    assertThat(endState).isEqualTo(MessageFrame.State.COMPLETED_SUCCESS);
+  }
+
+  @Test
+  public void sadPaths() {
+    try (InputStream failVectors =
+        KZGPointEvalPrecompileContractTest.class.getResourceAsStream("fail_pointEvaluation.json")) {
```
Comment (siladu) [inline evm/src/test/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompileContractTest.java:64]: like this approach, seems nice and straight forward

File: evm/src/test/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompileContractTest.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      JsonNode failJson = jsonMapper.readTree(failVectors);
+
+      for (JsonNode testCase : failJson) {
+
+        Bytes input = Bytes.fromHexString(testCase.get("Input").asText());
+
+        PrecompiledContract.PrecompileContractResult result =
+            contract.computePrecompile(input, toRun);
+        MessageFrame.State endState = result.getState();
+        assertThat(endState).isEqualTo(MessageFrame.State.COMPLETED_FAILED);
```
Comment (siladu) [inline evm/src/test/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompileContractTest.java:76]: I was expecting the 'ExpectedError' field would be asserted, is that not possible here?

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java [hunk 7]
```diff
@@ -86,4 +88,11 @@ public static void populateForBLS12(
...
+    // TODO: allow override to be configured?
```
Comment (siladu) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java:95]: Is this intended to be done in another PR?

**Review by siladu — COMMENTED** [review]
Added some comments but still want to understand and play with the tests a bit more.

Regarding the task list on #4822...
> * package the mainnet kzg trusted setup file so it is delivered via distribution in such a way that it can be loaded by native lib from the filesystem; loading from classpath won't work.

What was the solution for this?

> * add support for reftests that cover it

Was this done in this PR?

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java (hunk 4 above)
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:76]: Nope, everything comes out of the native lib as a generic RuntimeException, i'm just eating this one silently.  I didn't want to support re-loading of the setup after it is set.

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java (hunk 7 above)
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java:95]: Yes, I haven't decided if we should allow user-specified trusted setups, am loosely in favor of it at some point, but isn't needed yet.

File: evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java [hunk 8]
```diff
@@ -15,19 +15,25 @@
...
+    this.kzgTrustedSetupPath = pathToKZGTrustedSetup;
   }
 
   public long getJumpDestCacheWeightBytes() {
     return jumpDestCacheWeightKB * 1024L;
   }
 
-  public long getJumpDestCacheWeightKB() {
-    return jumpDestCacheWeightKB;
+  public Optional<String> getKzgTrustedSetupPath() {
```
[the remark this answers is not in the document]
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java:36]: It isn't yet, but if we decide to allow users to specify their own setup file, this will be called.

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java (hunk 2 above)
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/precompile/MainnetPrecompiledContracts.java:94]: Only because shipping fork for BLS12 is still tbd, so didn't want to build on top of that.

**Comment by jflo** [comment]
> 

Loading from classpath is fine, and user will be able to override that later.  We are still waiting on retesteth updates to cover it.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetPrecompiledContractRegistries.java [hunk 9]
```diff
@@ -60,6 +61,13 @@ public static PrecompileContractRegistry bls12(
     return registry;
   }
 
+  public static PrecompileContractRegistry cancun(
```
Comment (fab-10) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetPrecompiledContractRegistries.java:64]: this makes sure that this precompile is only seens when Cancun activates, right?

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java [hunk 10]
```diff
@@ -44,9 +46,18 @@ public static EvmOptions create() {
   private Long jumpDestCacheWeightKilobytes =
       32_000L; // 10k contracts, (25k max contract size / 8 bit) + 32byte hash
 
+  @SuppressWarnings({"FieldCanBeFinal", "FieldMayBeFinal"})
+  @CommandLine.Option(
+      names = {KZG_TRUSTED_SETUP_FILE_PATH},
+      description =
+          "absolute file path on the host system which points to the trusted KZG setup file to use",
+      arity = "1")
+  private String kzgTrustedSetupFilePath = null;
```
Comment (fab-10) [inline besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java:55]: Could this be `Path` instead of `String` like is done for other similar options that point to files? this will also help in fail fast in case the argument does not point to an existing file

File: evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java [hunk 11]
```diff
@@ -15,19 +15,25 @@
...
 
+import java.util.Optional;
+
 public class EvmConfiguration {
-  public static final EvmConfiguration DEFAULT = new EvmConfiguration(32_000L);
+  public static final EvmConfiguration DEFAULT = new EvmConfiguration(32_000L, Optional.empty());
   private final long jumpDestCacheWeightKB;
 
-  public EvmConfiguration(final long jumpDestCacheWeightKB) {
+  private final Optional<String> kzgTrustedSetupPath;
```
Comment (fab-10) [inline evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java:24]: as for previous comment, prefer `Path` type for the new field

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        File jniWillLoadFrom = File.createTempFile("kzgTrustedSetup", "txt");
+        Files.copy(is, jniWillLoadFrom.toPath(), REPLACE_EXISTING);
+        is.close();
+        pathToSetup = jniWillLoadFrom.getAbsolutePath();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    try {
+      BufferedReader setupFile =
```
Comment (fab-10) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:59]: convert to try-with-resources?

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+      final Bytes input, @NotNull final MessageFrame messageFrame) {
+
+    if (input.size() != 192) {
+      return new PrecompileContractResult(
+          Bytes.EMPTY,
+          false,
+          MessageFrame.State.COMPLETED_FAILED,
+          Optional.of(ExceptionalHaltReason.PRECOMPILE_ERROR));
+    }
+    // Bytes versionedHash = input.slice(0, 32);
```
Comment (fab-10) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:113]: this should be used to assert the below part, right? Will it be implemented in another PR?
> 
>     # Verify commitment matches versioned_hash
>     assert kzg_to_versioned_hash(commitment) == versioned_hash

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    Bytes output = Bytes.EMPTY;
+    PrecompileContractResult result;
+    try {
+      boolean proved =
+          CKZG4844JNI.verifyKzgProof(
+              commitment.toArray(), z.toArray(), y.toArray(), proof.toArray());
+
+      // # Return FIELD_ELEMENTS_PER_BLOB and BLS_MODULUS as padded 32 byte big endian values
+      //    return Bytes(U256(FIELD_ELEMENTS_PER_BLOB).to_be_bytes32() +
+      // U256(BLS_MODULUS).to_be_bytes32())
```
Comment (fab-10) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:128]: Is this commented because it will be completed later?

**Review by fab-10 — COMMENTED** [review]
looks good, please check the comments

File: CHANGELOG.md [hunk 15]
```diff
@@ -9,6 +9,7 @@
 ### Additions and Improvements
 - Added option to evm CLI tool to allow code execution at specific forks [#4913](https://github.com/hyperledger/besu/pull/4913)
 - Improve get account performance by using the world state updater cache [#4897](https://github.com/hyperledger/besu/pull/4897)
+- Add new KZG precompile and option to override the trusted setup being used.
```
Comment (fab-10) [inline CHANGELOG.md:12]: add link

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java (hunk 10 above)
Comment (fab-10) [reply besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java:62]: bump

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java [hunk 16]
```diff
@@ -733,17 +733,7 @@ static ProtocolSpecBuilder cancunDefinition(
...
-        .contractCreationProcessorBuilder(
-            (gasCalculator, evm) ->
-                new ContractCreationProcessor(
-                    gasCalculator,
-                    evm,
-                    true,
-                    List.of(
-                        MaxCodeSizeRule.of(contractSizeLimit), EOFValidationCodeRule.of(1, false)),
-                    1,
-                    SPURIOUS_DRAGON_FORCE_DELETE_WHEN_EMPTY_ADDRESSES))
```
Comment (fab-10) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java:746]: is this code removal intended?

File: evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java (hunk 11 above)
Comment (fab-10) [reply evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java:24]: bump

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    String pathToSetup;
+    CKZG4844JNI.Preset bitLength;
+    if (pathToTrustedSetup.isPresent()) {
+      pathToSetup = pathToTrustedSetup.get();
+    } else {
+      InputStream is =
+          KZGPointEvalPrecompiledContract.class.getResourceAsStream(
+              "mainnet_kzg_trusted_setup_4096.txt");
+      try {
+        File jniWillLoadFrom = File.createTempFile("kzgTrustedSetup", "txt");
```
Comment (fab-10) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:50]: you can set delete on exit on this temp file, so it is automatically cleaned

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java (hunk 13 above)
Comment (fab-10) [reply evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:113]: bump

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java (hunk 14 above)
Comment (fab-10) [reply evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:128]: bump

**Review by fab-10 — COMMENTED** [review]
Some old and new comments

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java (hunk 14 above)
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:128]: i left that in because it is from the spec and explains what is going on.

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java (hunk 17 above)
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:50]: I thought createTempFile did that automatically, thanks.

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java [hunk 18]
```diff
@@ -65,7 +66,10 @@ public static EvmOptions create() {
   @Override
   public EvmConfiguration toDomainObject() {
     return new EvmConfiguration(
-        jumpDestCacheWeightKilobytes, Optional.ofNullable(kzgTrustedSetupFilePath));
+        jumpDestCacheWeightKilobytes,
+        kzgTrustedSetupFilePath != null
+            ? Optional.ofNullable(Path.of(kzgTrustedSetupFilePath))
```
Comment (fab-10) [inline besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java:71]: nit: result of Path::of is never null

**Review by fab-10 — APPROVED** [review]
LGTM, just a nit

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java [hunk 19]
```diff
@@ -29,6 +31,8 @@ public class EvmOptions implements CLIOptions<EvmConfiguration> {
   /** The constant JUMPDEST_CACHE_WEIGHT. */
   public static final String JUMPDEST_CACHE_WEIGHT = "--Xevm-jumpdest-cache-weight-kb";
 
+  public static final String KZG_TRUSTED_SETUP_FILE_PATH = "--evm-kzg-trusted-setup-path";
```
Comment (shemnon) [inline besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java:34]: I think this should be an `--X` option.  Once the ceremony is done we will have the kzg embedded in the distribution, correct?  So this is just transitional for testing?

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java (hunk 18 above)
Comment (shemnon) [reply besu/src/main/java/org/hyperledger/besu/cli/options/unstable/EvmOptions.java:71]: This is not a base level configuration option.  This is the wrong place in code to do it.  It should be an argument in the MainnetEVMs methods for relevant versions.

File: build.gradle [hunk 20]
```diff
@@ -109,7 +109,7 @@ allprojects {
     }
     maven {
       url 'https://artifacts.consensys.net/public/maven/maven/'
-      content { includeGroupByRegex('tech\\.pegasys\\..*') }
+      content { includeGroupByRegex('tech\\.pegasys.*') }
```
Comment (shemnon) [inline build.gradle:112]:
```suggestion
      content { includeGroupByRegex('tech\\.pegasys(\\..*)?') }
```

Removing the dot opens the door wide to uncontrolled DNS names like `tech.pegasys-but-not-really:comprimised-kzg`.  Instead it should be listed as an optional section.

File: datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java [hunk 21]
```diff
@@ -71,7 +71,8 @@ public class Address extends DelegatingBytes implements org.hyperledger.besu.plu
   /** The constant BLS12_MAP_FP2_TO_G2. */
   public static final Address BLS12_MAP_FP2_TO_G2 = Address.precompiled(0x12);
 
-  /** The constant ZERO. */
+  public static final Address KZG_POINT_EVAL = Address.precompiled(0x14);
```
Comment (shemnon) [inline datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java:74]: Is this address fixed already?  Has BLS been "locked in" to those addresses?

File: ethereum/core/build.gradle [hunk 22]
```diff
@@ -15,6 +15,10 @@
 
 apply plugin: 'java-library'
 
+repositories {
+  maven { url "https://artifacts.consensys.net/public/maven/maven/" }
+}
+
```
Comment (shemnon) [inline ethereum/core/build.gradle:21]: This should not be needed here and should be removed.  The repository is already enumerated in the parent build file.

File: ethereum/core/build.gradle [hunk 23]
```diff
@@ -54,6 +58,7 @@ dependencies {
   implementation 'org.apache.tuweni:tuweni-concurrent'
   implementation 'org.apache.tuweni:tuweni-units'
   implementation 'org.apache.tuweni:tuweni-rlp'
+  implementation 'tech.pegasys:jc-kzg-4844:0.1.0'
```
Comment (shemnon) [inline ethereum/core/build.gradle:61]: Versions should not be listed in child build files.  Please add an entry to `<root>/gradle/versions.gradle`

File: evm/build.gradle [hunk 24]
```diff
@@ -30,6 +30,9 @@ jar {
   }
 }
 
+repositories {
+  maven { url "https://artifacts.consensys.net/public/maven/maven/" }
+}
```
Comment (shemnon) [inline evm/build.gradle:35]: This should not be here, the root gradle should cover this.  If it doesn't it should be fixed there.

File: evm/build.gradle [hunk 25]
```diff
@@ -43,9 +46,11 @@ dependencies {
   implementation 'org.apache.tuweni:tuweni-bytes'
   implementation 'org.hyperledger.besu:arithmetic'
   implementation 'org.hyperledger.besu:bls12-381'
+  implementation 'tech.pegasys:jc-kzg-4844:0.1.0'
```
Comment (shemnon) [inline evm/build.gradle:49]: versions should not be in module build files.  Enumerate them in `versions.gradle`

File: evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java [hunk 26]
```diff
@@ -15,20 +15,19 @@
...
+  public static final EvmConfiguration DEFAULT = new EvmConfiguration(32_000L, Optional.empty());
   private final long jumpDestCacheWeightKB;
 
-  /**
-   * Instantiates a new Evm configuration.
-   *
-   * @param jumpDestCacheWeightKB the jump dest cache weight kb
-   */
-  public EvmConfiguration(final long jumpDestCacheWeightKB) {
+  private final Optional<Path> kzgTrustedSetupPath;
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java:25]: The KZG setup is only relevant to some versions of the EVM, so it is not appropriate as a global configuration option.  Treat it like `ChainID` has been treated.

**Review by shemnon — CHANGES_REQUESTED** [review]
I don't like how the KZG files are being brought in.  Especially if we already have hard-coded a default as part of the library.  It's a config for one precompile so it shouldn't impact the configuration of the base EVM.

Also, once it's up and running how often will alternate KZGs be brought in?  Do we really need configurability?

**Comment by jflo** [comment]
> I don't like how the KZG files are being brought in. Especially if we already have hard-coded a default as part of the library. It's a config for one precompile so it shouldn't impact the configuration of the base EVM.
> 
> Also, once it's up and running how often will alternate KZGs be brought in? Do we really need configurability?

Yeah, the cart may be before the horse. I'm gonna remove configurability for interop, and we can add it back in if we actually discover a use case for it.

File: datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java (hunk 21 above)
Comment (jflo) [reply datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java:74]: As far as i know it is, the specs have been stable on it.

File: evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java [hunk 27]
```diff
@@ -38,6 +37,10 @@
 
 public class KZGPointEvalPrecompiledContract implements PrecompiledContract {
 
+  public KZGPointEvalPrecompiledContract() {
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/precompile/KZGPointEvalPrecompiledContract.java:40]: Needs Javadoc

**Review by shemnon — APPROVED** [review]
